### PR TITLE
Consensus: Remove ISM

### PR DESCRIPTION
**Pull request by NicolasDorier**
I hard coded ISM SF in regtest and adapted the tests.
Will try a sync on mainnet and testnet soon.

Note that even if the activation threshold (75% which enforce a rule for block version above a number) and enforcement threshold (95% reject old versions) are different, only the enforcement one really matter post fork.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -2374,13 +2369,13 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
     // when 75% of the network has upgraded:
-    if (block.nVersion >= 3 && IsSuperMajority(3, pindex->pprev, chainparams.GetConsensus().nMajorityEnforceBlockUpgrade, chainparams.GetConsensus())) {
+    if (block.nVersion >= 3 && pindex->nHeight >= chainparams.GetConsensus().BIP66Height) {
         flags |= SCRIPT_VERIFY_DERSIG;
     }
 
     // Start enforcing CHECKLOCKTIMEVERIFY, (BIP65) for block.nVersion=4
     // blocks, when 75% of the network has upgraded:
-    if (block.nVersion >= 4 && IsSuperMajority(4, pindex->pprev, chainparams.GetConsensus().nMajorityEnforceBlockUpgrade, chainparams.GetConsensus())) {
+    if (block.nVersion >= 4 && pindex->nHeight >= chainparams.GetConsensus().BIP65Height) {
```
Comment (NicolasDorier) [inline src/main.cpp]: actually, I don't need to compare the version. BIP65Height represent the 95% threshold height, so the first condition is always true.

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -817,22 +817,23 @@ UniValue verifychain(const UniValue& params, bool fHelp)
...
-    for (int i = 0; i < consensusParams.nMajorityWindow && pstart != NULL; i++)
+    UniValue rv(UniValue::VOBJ);
+    bool activated = false;
+    switch(minVersion)
     {
-        if (pstart->nVersion >= minVersion)
-            ++nFound;
-        pstart = pstart->pprev;
+        case 2:
+            activated = consensusParams.BIP34Height >= pindex->nHeight;
```
Comment (laanwj) [inline src/rpc/blockchain.cpp]: Shouldn't this comparison be the other way around?

**Comment by NicolasDorier** [comment]
* Fixed the inverted condition of activation of ISM soft fork in RPC
* TestNet synched successfully from scratch

Still waiting for mainnet to finish... taking a while :(

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -817,22 +817,23 @@ UniValue verifychain(const UniValue& params, bool fHelp)
 }
 
 /** Implementation of IsSuperMajority with better feedback */
-static UniValue SoftForkMajorityDesc(int minVersion, CBlockIndex* pindex, int nRequired, const Consensus::Params& consensusParams)
+static UniValue SoftForkMajorityDesc(int minVersion, CBlockIndex* pindex, const Consensus::Params& consensusParams)
```
Comment (instagibbs) [inline src/rpc/blockchain.cpp]: "minVersion" is a bit confusing after this.

File: src/chainparams.cpp [hunk 4]
```diff
@@ -247,11 +245,10 @@ class CRegTestParams : public CChainParams {
     CRegTestParams() {
         strNetworkID = "regtest";
         consensus.nSubsidyHalvingInterval = 150;
-        consensus.nMajorityEnforceBlockUpgrade = 750;
-        consensus.nMajorityRejectBlockOutdated = 950;
-        consensus.nMajorityWindow = 1000;
-        consensus.BIP34Height = -1; // BIP34 has not necessarily activated on regtest
+        consensus.BIP34Height = 100000000; // BIP34 has not necessarily activated on regtest
         consensus.BIP34Hash = uint256();
+        consensus.BIP65Height = 1351; // BIP65 activated on regtest
```
Comment (instagibbs) [inline src/chainparams.cpp]: seem like magic numbers without explanation

File: src/main.cpp [hunk 5]
```diff
@@ -2372,15 +2367,13 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
 
     unsigned int flags = fStrictPayToScriptHash ? SCRIPT_VERIFY_P2SH : SCRIPT_VERIFY_NONE;
 
-    // Start enforcing the DERSIG (BIP66) rules, for block.nVersion=3 blocks,
-    // when 75% of the network has upgraded:
-    if (block.nVersion >= 3 && IsSuperMajority(3, pindex->pprev, chainparams.GetConsensus().nMajorityEnforceBlockUpgrade, chainparams.GetConsensus())) {
+    // Start enforcing the DERSIG (BIP66) rule
+    if (pindex->nHeight >= chainparams.GetConsensus().BIP66Height) {
```
Comment (instagibbs) [inline src/main.cpp]: nit: Any reason not to leave version check in? The fewer differences the better, imo.

File: src/chainparams.cpp [hunk 6]
```diff
@@ -167,11 +166,10 @@ class CTestNetParams : public CChainParams {
     CTestNetParams() {
         strNetworkID = "test";
         consensus.nSubsidyHalvingInterval = 210000;
-        consensus.nMajorityEnforceBlockUpgrade = 51;
-        consensus.nMajorityRejectBlockOutdated = 75;
-        consensus.nMajorityWindow = 100;
         consensus.BIP34Height = 21111;
         consensus.BIP34Hash = uint256S("0x0000000023b3a96d3484e5abb3755c413e7d41500f8e2a5c3f0dd01299cd8ef8");
+        consensus.BIP65Height = 581885; // 00000000007f6655f22f98e72ed80d8b06dc761d5da09df0fa1dc4be4f861eb6
+        consensus.BIP66Height = 330776; // 000000002104c8c45e99a8853285a3b592602a3ccde2b832481da85e9e4ba182
```
Comment (MarcoFalke) [inline src/chainparams.cpp]: Why is it not possible to check the hash for main and test?

File: src/rpc/blockchain.cpp [hunk 7]
```diff
@@ -897,13 +897,9 @@ UniValue getblockchaininfo(const UniValue& params, bool fHelp)
             "     {\n"
             "        \"id\": \"xxxx\",        (string) name of softfork\n"
             "        \"version\": xx,         (numeric) block version\n"
-            "        \"enforce\": {           (object) progress toward enforcing the softfork rules for new-version blocks\n"
+            "        \"reject\": {            (object) progress toward rejecting pre-softfork blocks (same fields as \"enforce\")\n"
```
Comment (instagibbs) [inline src/rpc/blockchain.cpp]: This comment doesn't really make sense to me since "enforce" is no longer anywhere.

File: src/chainparams.cpp [hunk 8]
```diff
@@ -247,11 +245,10 @@ class CRegTestParams : public CChainParams {
     CRegTestParams() {
         strNetworkID = "regtest";
         consensus.nSubsidyHalvingInterval = 150;
-        consensus.nMajorityEnforceBlockUpgrade = 750;
-        consensus.nMajorityRejectBlockOutdated = 950;
-        consensus.nMajorityWindow = 1000;
-        consensus.BIP34Height = -1; // BIP34 has not necessarily activated on regtest
+        consensus.BIP34Height = 100000000; // BIP34 has not necessarily activated on regtest
```
Comment (MarcoFalke) [inline src/chainparams.cpp]: Why is this necessary?

File: src/chainparams.cpp (hunk 6 above)
Comment (instagibbs) [inline src/chainparams.cpp]: Where would that be useful? The hash is only used as an exception to BIP30 rule at https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L2358

File: src/chainparams.cpp (hunk 4 above)
Comment (NicolasDorier) [inline src/chainparams.cpp]: I'll update the comment, this magic number is chosen such that bip65 tests in python extended tests pass.

File: src/main.cpp (hunk 5 above)
Comment (NicolasDorier) [inline src/main.cpp]: Yes, will later make refactoring easier because this does not depend on Block anymore.
Moreover, it is impossible that `block.nVersion` to be below 3 at this point. I'm using the 95% threshold for this test, if the `block.nVersion` were below 3 it would have been rejected before by https://github.com/bitcoin/bitcoin/pull/8391/files#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR3516

File: src/rpc/blockchain.cpp (hunk 3 above)
Comment (NicolasDorier) [inline src/rpc/blockchain.cpp]: What would be better ?

File: src/main.cpp (hunk 5 above)
Comment (sipa) [inline src/main.cpp]: The IsSuperMajority call is slow.

File: src/rpc/blockchain.cpp (hunk 3 above)
Comment (instagibbs) [inline src/rpc/blockchain.cpp]: "version", like the caller SoftForkDesc

File: src/main.cpp (hunk 5 above)
Comment (instagibbs) [inline src/main.cpp]: @sipa? Hm? I was just asking about nVersion check.  

That said, future refactoring purposes is good enough for me.

File: src/consensus/params.h [hunk 9]
```diff
@@ -39,13 +39,13 @@ struct BIP9Deployment {
...
     uint256 hashGenesisBlock;
     int nSubsidyHalvingInterval;
-    /** Used to check majorities for block version upgrade */
-    int nMajorityEnforceBlockUpgrade;
-    int nMajorityRejectBlockOutdated;
-    int nMajorityWindow;
     /** Block height and hash at which BIP34 becomes active */
     int BIP34Height;
     uint256 BIP34Hash;
+    /** Block height at which BIP65 becomes active */
```
Comment (jtimon) [inline src/consensus/params.h]: We can make an array with an enum, like with vDeployments. Say, vPastDeployments, vOldDeployments, vBuriedDeployments, or something of the sort.

File: src/main.cpp (hunk 5 above)
Comment (jtimon) [inline src/main.cpp]: Yeah, as @NicolasDorier points out the block.nVersion check can be safely removed from here because blocks with invalid versions will be rejected in ContextualCheckBlockHeader.

File: src/consensus/params.h (hunk 9 above)
Comment (NicolasDorier) [inline src/consensus/params.h]: I prefer keeping change of logic and refactoring done in different PR. I want this PR to be easy to review, so I'm using the previous way of doing with BIP34Height.

Comment (jtimon) [inline src/consensus/params.h]: Well, you can leave BIP34 as it is and put the new things in the array if you think that is less disruptive, no?
Anyway, if you feel strongly against doing it now, we can do it later. It's just feel cleaner not to put code to destroy it "right afterwards" (I can foresee the same "we can do the refactor later" argument when moving BIP113 there...).

File: src/chainparams.cpp (hunk 6 above)
Comment (jtimon) [inline src/chainparams.cpp]: Well, in fact, what that's doing is that if we're past BIP34Height but at BIP34Height we don't have the hardcoded block, then keep checking bip30 as if bip34 hadn't been activated. 
But now in https://github.com/bitcoin/bitcoin/pull/8391/files#diff-7ec3c68a81efff79b6ca22ac1f1eabbaL3550 if we have a reorg below BIP34Height, BIP34 will be activated after BIP34Height regardless of the original activation being different, thus the optimization of not checking BIP30 if BIP34 is activated could be used as well. So BIP34Hash can go after this change, I believe.
If we have a reorg to BIP34Height things are terribly broken anyway.

**Comment by jtimon** [comment]
utACK besides nits.

File: src/chainparams.cpp (hunk 8 above)
Comment (NicolasDorier) [inline src/chainparams.cpp]: I just re added it with this check, tests with block v1 on regnet get rejected by the "obsolete version" check like  https://travis-ci.org/bitcoin/bitcoin/jobs/146661777 .  It makes sense as if BIP34Height is -1 then https://github.com/bitcoin/bitcoin/pull/8391/files#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR3515 will kick out.

**Comment by NicolasDorier** [comment]
* BIP34 does not activate on regnet. (https://github.com/bitcoin/bitcoin/pull/8391/files#diff-64cbe1ad5465e13bc59ee8bb6f3de2e7R248), rationale explained on https://github.com/bitcoin/bitcoin/pull/8391#discussion_r71903685

**Comment by NicolasDorier** [comment]
Sync of Testnet succeed, the Travis error is unrelated to this PR is it possible to run it again ?

**Comment by NicolasDorier** [comment]
Intial sync of testnet and mainnet succeed.

**Comment by jtimon** [comment]
Here are my proposed Here are my proposed changes in code: https://github.com/jtimon/bitcoin/compare/remove-ism...consensus-post-remove-ism

In any case, utACK 122786d

**Comment by sipa** [comment]
Note that #8418 adds a means to override the activation times for BIP9 activations in regtest mode. Perhaps a compatible approach can be followed here? Or is there no need for testing with different softforks activate in regtest?

**Comment by sipa** [comment]
utACK 122786d0e0170c73536360b705af711e1338adbf

**Comment by NicolasDorier** [comment]
I'll do a separate PR on top of this one for using the approach of #8418 alongs with @jtimon suggestion.
I prefer keeping this PR as simple as I can.

**Comment by laanwj** [comment]
ACK 122786d

**Comment by sdaftuar** [comment]
@morcos and I were discussing this, and I think this change is subtle enough that it deserves some kind of BIP write-up.

Technically this change is a hardfork: old ISM-based code, when presented with an alternate chain history in which one of these softforks activated earlier, would then reject a subsequent block that violated the softfork before the hardcoded heights here, whereas this code would not reject such a block (since the softfork would not yet be enforced).  This seems like a fundamental consequence of removing the ISM code (excluding doing something like checkpointing the whole headers chain to detect this kind of thing, which seems like obvious overkill).

In my judgement this is fine, as by the time we deploy this code in 0.14, it'll have been over a year since the most recent one of these softforks deployed.  But to further everyone's understanding of how these kinds of consensus changes are made (especially for future developers, as well as authors of alternate implementations) I think it'd be beneficial to document this change along with the considerations and rationale that went into it.

**Comment by jtimon** [comment]
I agree a BIP makes sense.

**Comment by NicolasDorier** [comment]
will write BIP
